### ISSUE TITLE
d*LR report scales throughout training

### ISSUE THREAD

**Issue by rockerBOO** [LoganBooker/prodigy-plus-schedule-free#23]
This behavior was happening for with/without schedulefree on 1.9.0 and the 2.0.0 dev3. 

It is happening when I am using LoRA plus which makes 2 parameter groups. You can see the "plus" is normal but the non-plus is scaling. It will keep scaling until the end of training where I had it up to 1e+13 or more. It isn't effecting the actual training though as it is more relative to what is being reported in group1 (unet plus).

So group0 scales but group 1 does not. In the reporting. I could try to make an example script to showcase this without training a model but reporting this as is. 

![Image](https://github.com/user-attachments/assets/806964bc-e174-4f1e-85c2-5391d68595bf)

**Comment by LoganBooker** [comment]
Hi rockerBOO, I've encountered this behaviour as well with LoRA-FA (https://arxiv.org/abs/2308.03303), though I've also seen it happen with SDXL LoRAs more generally when training the TEs and UNet as separate groups. The UNet seems to struggle to find an LR, and just grows forever instead. This behaviour is masked when training as a single group, as the `d` is balanced out.

I've tried various ways of combatting it, but they all have drawbacks. You can try disabling StableAdamW -- my theory is that the RMS normalisation dampens the initial updates (which can have and RMS of ~9 to ~31 depending on `beta2`) and cripples the UNet's ability to escape zero-initialisation. On 2.0.0 dev3 with StableAdamW disabled, the optimiser is effectively using the reference Prodigy logic (just with second moment factorisation), so if that also fails, then it's a weakness of Prodigy in this particular scenario.

I'll keep experimenting of course, but this is one of the reasons I implemented `prodigy_steps` so you'd have some way to stop Prodigy after a certain point as a failsafe.

**Comment by rockerBOO** [comment]
So it is actually increasing the `d` and not just a visual bug? It seems to train fine and not overtrain so it seems odd. If it was doing 1e+13 i'd imagine it would have some immense issues though? Are you sure it's not just some visual bug in this case?

**Comment by LoganBooker** [comment]
Ah apologies, looks like I misunderstood. :P The optimiser doesn't do anything weird with `d` that would make the reporting show inflated values like that, but I can take a look.

**Comment by rockerBOO** [comment]
I can NOT get it to run with the following script using the same settings and format as I'm doing to train a LoRA. Maybe something is going on in the reporting so I will look at it a little deeper to see if it's some implementation issue. If I can isolate it I will modify the following script to try to replicate it. If you see anything that you'd like to see in the following script let me know. 

```python
from prodigyplus import ProdigyPlus
import torch
import torch.nn.functional as F


class TestModule(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.lin = torch.nn.Linear(10, 50)

    def forward(self, x):
        x = self.lin(x)
        return x


# Set random seed for reproducibility
torch.manual_seed(42)

# Create modules
module = TestModule()
module_2 = TestModule()

module.train()
module_2.train()


# Create optimizer with more reasonable learning rates
parameters = [{"params": module.parameters(), "lr": 0.01}, {"params": module_2.parameters(), "lr": 0.05}]

# Initialize the optimizer
optimizer = ProdigyPlus(parameters, eps=None, use_speed=True, use_adopt=True, weight_decay=1, d_coef=1)

# Create a fixed reference target
target = torch.randn(1, 50) * 10

optimizer.train()

# Training loop
for i in range(100000):
    # Use fixed input instead of random each time
    x = torch.randn(1, 10)

    # Forward pass
    a = module(x)
    b = module_2(x)

    # Use separate loss functions that both minimize
    loss1 = F.mse_loss(a, target)
    loss2 = F.mse_loss(b, target)

    # Simple combined loss (both minimizing)
    loss = loss1 + loss2

    # Zero gradients, backward pass, and step
    optimizer.zero_grad()  # Add this line to clear previous gradients
    loss.backward()
    optimizer.step()

    # Print diagnostics every 100 iterations
    if i % 100 == 0:
        print(f"Iteration {i}")
        print(f"Loss 1: {loss1.item():.6f}, Loss 2: {loss2.item():.6f}")

        # Print learning rate information
        for j, group in enumerate(optimizer.param_groups):
            dlr = optimizer.get_dlr(group)
            print(f"Group {j}, DLR: {dlr:.6f}")

        # Print some parameter statistics to check for explosion
        for j, group in enumerate(optimizer.param_groups):
            params = next(iter(group["params"]))
            print(f"Group {j} param stats: Mean={params.mean().item():.6f}, Std={params.std().item():.6f}")
        print("-" * 50)
```

**Comment by LoganBooker** [comment]
Thank rockerBOO, this helped quite a bit. It actually isn't misreporting and there are a few things going on which made it a bit tricker to figure out.

**Short version**
SPEED should be used with `weight_decay=0` (or very small values, like 0.01 or less), and with `split_groups_mean=False`. I'll either need to enforce these within the optimiser when SPEED is enabled, or add different logic.

**Long version**
The optimiser settings in the test script don't have `split_groups_mean=False`, so the DLR is calculated using the harmonic mean of all the `d` values multiplied by the individual LRs of the groups. This means that if one group's `d` explodes, but the other doesn't, it'll be "balanced out" to a more rational `d`. It also means if you look at the per-group `d` values, it's possible to see some extremes that don't reflect the real LR (which is what the graphs in your original post show).

On top of this, SPEED doesn't seem to play well with weight decay, and will quickly explode with values greater than ~0.01.

So everything is being reported correctly, it's just that the final LR used is sane because the harmonic mean of `d` values is being used, so the outlier is cancelled out to some degree.

I think I might change the default value of `split_group_mean` to `False` for the final 2.0.0 release, so the optimiser uses per-group `d * lr`.

Thanks for digging into this, it is extremely helpful!

**Comment by rockerBOO** [comment]
In my actual usage in Kohya i'm also not setting `split_groups_mean=False` so it defaults to True? Wouldn't explain why it is exploding the value like that. It also isn't effecting the end result negatively... I will take what you mentioned and try some things. Maybe the high weight decay was hiding some things so I will see what happens there.

**Comment by LoganBooker** [comment]
split_groups_mean is set to True by default. It's not responsible for the exploding values, weight decay of 1 is. split_groups_mean just hides the issue.
